cpp/src/arrow/type_traits.h

```diff
@@ -435,6 +435,8 @@ struct TypeTraits<DenseUnionType> {
   using ArrayType = DenseUnionArray;
   using BuilderType = DenseUnionBuilder;
   using ScalarType = DenseUnionScalar;
+  using ValueOffsetBuilderType = Int32Builder;
+  using ChildIdBuilderType = Int8Builder;
```

Comment:
These are not generally useful traits since they're not shared by other types. Please revert them